File: ios/Jamboree/Games/President/PresidentEngine.swift
import Foundation

/// Standard 52-card deck used by President. Distinct from `CrazyEights` /
/// `Cheat` so the engines stay independent. Card order low→high here is
///   3 4 5 6 7 8 9 10 J Q K A 2
/// with twos as the highest rank, so we store ranks as 2..14 and re-rank
/// in `power(_:)`.
enum PresSuit: String, CaseIterable {
    case clubs, diamonds, hearts, spades
    var short: String {
        switch self { case .clubs: "C"; case .diamonds: "D"; case .hearts: "H"; case .spades: "S" }
    }
    var glyph: String {
        switch self { case .clubs: "♣"; case .diamonds: "♦"; case .hearts: "♥"; case .spades: "♠" }
    }
    var isRed: Bool { self == .diamonds || self == .hearts }
}

struct PresCard: Hashable {
    let suit: PresSuit
    /// 2..14 where J=11, Q=12, K=13, A=14, 2 stays 2.
    let rank: Int

    var rankShort: String {
        switch rank { case 11: "J"; case 12: "Q"; case 13: "K"; case 14: "A"; default: "\(rank)" }
    }
    var id: String { "\(suit.short)\(rank)" }
    var display: String { "\(rankShort)\(suit.glyph)" }

    /// Power for trick-comparison: 3 (lowest) … 2 (highest).
    /// 3..10 keep their face value (3..10); J..K → 11..13; A → 14; 2 → 15.
    var power: Int {
        if rank == 2 { return 15 }
        return rank
    }
}

func presStandardDeck() -> [PresCard] {
    PresSuit.allCases.flatMap { s in (2...14).map { PresCard(suit: s, rank: $0) } }
}

enum PresidentPhase { case lobby, swapping, playing, gameOver }

enum PresRank: String, CaseIterable {
    case president, vicePresident, viceScum, scum, neutral

    var label: String {
        switch self {
        case .president: "President"
        case .vicePresident: "Vice President"
        case .viceScum: "Vice Scum"
        case .scum: "Scum"
        case .neutral: "Neutral"
        }
    }
}

/// One play in the current trick. Combinations: singles, pairs, triples,
/// quads, or runs of pairs (consecutive ranks; length even).
enum PresCombo: Equatable {
    case single, pair, triple, quad, runOfPairs(length: Int)
}

struct PresOptions: Equatable {
    /// President can announce a "house rule" each round (free-text).
    var allowHouseRules: Bool = false
    /// Playing 4-of-a-kind inverts card power for the rest of the trick.
    /// (No-op in our pure engine — recorded for UI display only.)
    var revolution: Bool = false
}

final class PresidentPlayer {
    let id: String, name: String, isHost: Bool
    var hand: [PresCard] = []
    var rank: PresRank = .neutral
    /// True if this player finished the round (placed all cards).
    var finished: Bool = false
    /// 1-based finishing order (1 = president). 0 means not finished.
    var finishOrder: Int = 0
    init(id: String, name: String, isHost: Bool) { self.id = id; self.name = name; self.isHost = isHost }
}

/// The currently-open trick.
struct PresTrick {
    let combo: PresCombo
    let topPower: Int
    let leaderId: String
}

/// Pending card-swap action between rounds.
struct PresSwap {
    let fromId: String
    let toId: String
    /// Number of cards the `fromId` player must give.
    let count: Int
    /// True if the giver chooses (Pres back, VPs choose); false if it's
    /// the giver's "best" cards (Scum→Pres, VS→VP).
    let giverChooses: Bool
    /// Once selected by the giver, the actual cards.
    var cards: [PresCard]?
}

/// Pure President / Scum / Asshole engine.
final class PresidentEngine {
    private var rng: any RandomNumberGenerator
    private(set) var players: [String: PresidentPlayer] = [:]
    private var playerOrder: [String] = []
    private(set) var seating: [String] = []
    var phase: PresidentPhase = .lobby
    private(set) var options = PresOptions()

    /// Current trick descriptor, nil between tricks.
    var trick: PresTrick?
    /// Players (in seating order, by id) who have passed during the current trick.
    var passedThisTrick: Set<String> = []
    /// Last play descriptor for UI (cards just played + by whom).
    var lastPlay: (playerId: String, cards: [PresCard], combo: PresCombo)?
    /// Finishing order (player ids in order they emptied their hand).
    var finishOrder: [String] = []
    var currentIndex: Int = 0
    var lastEvent: String?
    var roundNumber: Int = 0
    var pendingSwaps: [PresSwap] = []

    let tutorialVote = TutorialVote()

    init(rng: any RandomNumberGenerator = SystemRandomNumberGenerator()) { self.rng = rng }

    var current: PresidentPlayer? { seating.isEmpty ? nil : players[seating[currentIndex]] }
    var canStart: Bool { phase == .lobby && (4...7).contains(players.count) }

    @discardableResult
    func addPlayer(id: String, name: String, isHost: Bool = false) -> PresidentPlayer {
        let p = PresidentPlayer(id: id, name: name, isHost: isHost)
        if players[id] == nil { playerOrder.append(id) }
        players[id] = p
        return p
    }
    func removePlayer(_ id: String) {
        if phase == .lobby { players[id] = nil; playerOrder.removeAll { $0 == id } }
    }

    func setOptions(_ o: PresOptions) {
        guard phase == .lobby else { return }
        options = o
    }

    /// Start the next round. On the first round, ranks haven't been
    /// assigned so we deal and the holder of 3♣ leads. Subsequent rounds
    /// open a swap phase before dealing-to-play.
    func start() {
        guard canStart else { return }
        seating = playerOrder.shuffled(using: &rng)
        for p in players.values { p.rank = .neutral; p.finished = false; p.finishOrder = 0; p.hand.removeAll() }
        finishOrder.removeAll()
        passedThisTrick.removeAll()
        trick = nil
        lastPlay = nil
        roundNumber = 1
        deal()
        // First round: 3-of-clubs leads.
        if let lead = seating.first(where: { id in players[id]!.hand.contains(PresCard(suit: .clubs, rank: 3)) }) {
            currentIndex = seating.firstIndex(of: lead)!
        } else {
            currentIndex = 0
        }
        phase = .playing
        lastEvent = "Round 1 — \(current!.name) leads with 3♣"
    }

    /// Begin a subsequent round: swap phase if ranks are set, else just deal+play.
    func startNextRound() {
        guard phase == .gameOver, !finishOrder.isEmpty else { return }
        assignRanks()
        roundNumber += 1
        for p in players.values { p.hand.removeAll(); p.finished = false; p.finishOrder = 0 }
        finishOrder.removeAll()
        trick = nil; lastPlay = nil; passedThisTrick.removeAll()
        deal()
        // Set up swap actions based on ranks.
        pendingSwaps = swapPlan()
        if pendingSwaps.isEmpty {
            // No swaps (e.g. 4-player game with no VPs — shouldn't happen)
            startTricksAfterSwaps()
        } else {
            phase = .swapping
            lastEvent = "Round \(roundNumber) — swap phase"
        }
    }

    /// Player provides a card selection for a pending swap. Returns nil on success.
    /// For "giver chooses" swaps the giver picks `count` cards; for "automatic
    /// best" swaps the engine ignores the input and uses the top cards.
    func submitSwap(fromId: String, cards: [PresCard]) -> String? {
        guard phase == .swapping else { return "not in swap phase" }
        guard let idx = pendingSwaps.firstIndex(where: { $0.fromId == fromId && $0.cards == nil })
        else { return "no pending swap for you" }
        let s = pendingSwaps[idx]
        let chosen: [PresCard]
        if s.giverChooses {
            guard cards.count == s.count else { return "give exactly \(s.count) card\(s.count == 1 ? "" : "s")" }
            chosen = cards
        } else {
            // Best == highest power.
            let sorted = players[fromId]!.hand.sorted { $0.power > $1.power }
            chosen = Array(sorted.prefix(s.count))
        }
        // Remove from giver, add to receiver.
        var giverHand = players[fromId]!.hand
        for c in chosen {
            guard let i = giverHand.firstIndex(of: c) else { return "card not in hand" }
            giverHand.remove(at: i)
        }
        players[fromId]!.hand = giverHand
        players[s.toId]!.hand.append(contentsOf: chosen)
        players[s.toId]!.hand.sort { $0.power < $1.power }
        players[fromId]!.hand.sort { $0.power < $1.power }
        pendingSwaps[idx].cards = chosen
        lastEvent = "\(players[fromId]!.name) gave \(chosen.count) card\(chosen.count == 1 ? "" : "s") to \(players[s.toId]!.name)"
        if pendingSwaps.allSatisfy({ $0.cards != nil }) {
            startTricksAfterSwaps()
        }
        return nil
    }

    private func startTricksAfterSwaps() {
        // President leads.
        if let pid = finishOrder.first, let i = seating.firstIndex(of: pid) {
            currentIndex = i
        } else {
            currentIndex = 0
        }
        phase = .playing
        lastEvent = "\(current!.name) (President) leads round \(roundNumber)"
        pendingSwaps.removeAll()
        finishOrder.removeAll()
        for p in players.values { p.finished = false; p.finishOrder = 0 }
    }

    /// Play a combination from `playerId`'s hand. Returns nil on success.
    func play(playerId: String, cards: [PresCard]) -> String? {
        guard phase == .playing else { return "not playing" }
        guard let p = players[playerId] else { return "unknown player" }
        guard p.id == current!.id else { return "not your turn" }
        guard !cards.isEmpty else { return "play at least one card" }
        // Must own all cards.
        var hand = p.hand
        for c in cards {
            guard let i = hand.firstIndex(of: c) else { return "card not in hand" }
            hand.remove(at: i)
        }
        guard let combo = classify(cards) else { return "invalid combination" }
        let power = comboPower(combo, cards: cards)
        if let t = trick {
            if !sameType(t.combo, combo) { return "must play \(describe(t.combo))" }
            if power <= t.topPower { return "must play higher than \(t.topPower)" }
            // First round constraint: opener must include 3♣ on round 1's first play.
        } else if roundNumber == 1 && lastPlay == nil
                  && p.hand.contains(PresCard(suit: .clubs, rank: 3))
                  && !cards.contains(PresCard(suit: .clubs, rank: 3)) {
            return "first play must include 3♣"
        }
        p.hand = hand
        trick = PresTrick(combo: combo, topPower: power, leaderId: trick?.leaderId ?? p.id)
        lastPlay = (p.id, cards, combo)
        passedThisTrick.removeAll() // Pass set tracks only players who passed since the last play.
        lastEvent = "\(p.name) played \(formatCards(cards)) (\(describe(combo)))"
        if p.hand.isEmpty {
            p.finished = true
            finishOrder.append(p.id)
            p.finishOrder = finishOrder.count
        }
        // If only one player still has cards, round ends.
        if remainingPlayers().count <= 1 {
            // Append any straggler.
            for sid in seating {
                if let pl = players[sid], !pl.finished {
                    pl.finished = true
                    finishOrder.append(pl.id)
                    pl.finishOrder = finishOrder.count
                }
            }
            assignRanks()
            phase = .gameOver
            lastEvent = "Round over"
            return nil
        }
        advanceTurn()
        return nil
    }

    /// Pass the current trick. Returns nil on success.
    func pass(playerId: String) -> String? {
        guard phase == .playing else { return "not playing" }
        guard let p = players[playerId] else { return "unknown player" }
        guard p.id == current!.id else { return "not your turn" }
        if trick == nil { return "lead — can't pass" }
        passedThisTrick.insert(p.id)
        lastEvent = "\(p.name) passed"
        // If every remaining player except the last leader has passed, the
        // trick closes and the last player to play leads.
        let alive = remainingPlayers().filter { !$0.finished }.map { $0.id }
        if let lp = lastPlay, alive.allSatisfy({ $0 == lp.playerId || passedThisTrick.contains($0) }) {
            // Trick closes.
            trick = nil
            passedThisTrick.removeAll()
            if let i = seating.firstIndex(of: lp.playerId), let player = players[lp.playerId], !player.finished {
                currentIndex = i
                lastEvent = "\(player.name) wins the trick and leads"
            } else {
                // Last player to play finished — next alive player leads.
                advanceTurn()
            }
            return nil
        }
        advanceTurn()
        return nil
    }

    func reset() {
        phase = .lobby
        for p in players.values { p.hand.removeAll(); p.rank = .neutral; p.finished = false; p.finishOrder = 0 }
        trick = nil; lastPlay = nil; passedThisTrick.removeAll()
        finishOrder.removeAll(); pendingSwaps.removeAll()
        currentIndex = 0; lastEvent = nil; roundNumber = 0
    }

    // MARK: - Internals

    private func deal() {
        let deck = presStandardDeck().shuffled(using: &rng)
        for (i, c) in deck.enumerated() {
            let pid = seating[i % seating.count]
            players[pid]!.hand.append(c)
        }
        for p in players.values { p.hand.sort { $0.power < $1.power } }
    }

    private func assignRanks() {
        guard !finishOrder.isEmpty else { return }
        let n = finishOrder.count
        for (i, id) in finishOrder.enumerated() {
            players[id]!.rank = rankFor(position: i, of: n)
            players[id]!.finishOrder = i + 1
        }
    }

    private func rankFor(position: Int, of total: Int) -> PresRank {
        if position == 0 { return .president }
        if position == total - 1 { return .scum }
        if total >= 4 && position == 1 { return .vicePresident }
        if total >= 4 && position == total - 2 { return .viceScum }
        return .neutral
    }

    /// Build the swap plan from current ranks. Scum→Pres = 2; VS→VP = 1.
    /// Both reverse exchanges are giverChooses (Pres picks back; VP picks back).
    private func swapPlan() -> [PresSwap] {
        var plan: [PresSwap] = []
        let pres = players.values.first { $0.rank == .president }
        let scum = players.values.first { $0.rank == .scum }
        let vp = players.values.first { $0.rank == .vicePresident }
        let vs = players.values.first { $0.rank == .viceScum }
        if let pres, let scum {
            plan.append(PresSwap(fromId: scum.id, toId: pres.id, count: 2,
                                  giverChooses: false, cards: nil))
            plan.append(PresSwap(fromId: pres.id, toId: scum.id, count: 2,
                                  giverChooses: true, cards: nil))
        }
        if let vp, let vs {
            plan.append(PresSwap(fromId: vs.id, toId: vp.id, count: 1,
                                  giverChooses: false, cards: nil))
            plan.append(PresSwap(fromId: vp.id, toId: vs.id, count: 1,
                                  giverChooses: true, cards: nil))
        }
        return plan
    }

    private func advanceTurn() {
        let n = seating.count
        for _ in 0..<n {
            currentIndex = (currentIndex + 1) % n
            let p = players[seating[currentIndex]]!
            if !p.finished { return }
        }
    }

    func remainingPlayers() -> [PresidentPlayer] {
        seating.compactMap { players[$0] }.filter { !$0.finished }
    }

    /// Classify a card group into a [PresCombo]. nil if not a legal play.
    func classify(_ cards: [PresCard]) -> PresCombo? {
        guard !cards.isEmpty else { return nil }
        let sorted = cards.sorted { $0.power < $1.power }
        let counts = Dictionary(grouping: sorted, by: { $0.rank }).mapValues { $0.count }
        if counts.count == 1 {
            switch sorted.count {
            case 1: return .single
            case 2: return .pair
            case 3: return .triple
            case 4: return .quad
            default: return nil
            }
        }
        // Run of pairs: every distinct rank has exactly 2 cards, ranks consecutive (by face value).
        if counts.values.allSatisfy({ $0 == 2 }) && counts.count >= 2 {
            let ranks = counts.keys.sorted()
            for i in 1..<ranks.count where ranks[i] != ranks[i-1] + 1 { return nil }
            return .runOfPairs(length: ranks.count)
        }
        return nil
    }

    /// Power for trick-beating: the highest power in the combo.
    func comboPower(_ combo: PresCombo, cards: [PresCard]) -> Int {
        cards.map { $0.power }.max() ?? 0
    }

    func sameType(_ a: PresCombo, _ b: PresCombo) -> Bool {
        switch (a, b) {
        case (.single, .single), (.pair, .pair), (.triple, .triple), (.quad, .quad):
            return true
        case let (.runOfPairs(la), .runOfPairs(lb)): return la == lb
        default: return false
        }
    }

    func describe(_ combo: PresCombo) -> String {
        switch combo {
        case .single: "single"
        case .pair: "pair"
        case .triple: "triple"
        case .quad: "four of a kind"
        case .runOfPairs(let len): "\(len) consecutive pairs"
        }
    }

    private func formatCards(_ cards: [PresCard]) -> String {
        cards.sorted { $0.power < $1.power }.map { $0.display }.joined(separator: " ")
    }
}

File: android/app/src/main/java/com/example/jamboree/games/president/PresidentEngine.kt
package com.example.ubapp.games.president

import kotlin.random.Random

/** Standard 52-card deck for President. Distinct from Crazy Eights / Cheat
 *  so the engines stay independent. Card power for trick-comparison: 3
 *  (lowest) → A → 2 (highest). */
enum class PresSuit(val short: String, val glyph: String) {
    CLUBS("C", "♣"), DIAMONDS("D", "♦"), HEARTS("H", "♥"), SPADES("S", "♠");
    val isRed: Boolean get() = this == DIAMONDS || this == HEARTS
}

/** rank 2..14 where 2 = two (highest), 11..13 = J/Q/K, 14 = A. */
data class PresCard(val suit: PresSuit, val rank: Int) {
    val rankShort: String get() = when (rank) {
        11 -> "J"; 12 -> "Q"; 13 -> "K"; 14 -> "A"; else -> "$rank"
    }
    val id: String get() = "${suit.short}$rank"
    override fun toString() = "$rankShort${suit.glyph}"
    /** Power: 3..14 keep value (3 low, A=14), 2 → 15 (top). */
    val power: Int get() = if (rank == 2) 15 else rank
}

fun presStandardDeck(): List<PresCard> =
    PresSuit.entries.flatMap { s -> (2..14).map { PresCard(s, it) } }

enum class PresidentPhase { LOBBY, SWAPPING, PLAYING, GAME_OVER }

enum class PresRank(val label: String) {
    PRESIDENT("President"),
    VICE_PRESIDENT("Vice President"),
    VICE_SCUM("Vice Scum"),
    SCUM("Scum"),
    NEUTRAL("Neutral");

    fun wireValue(): String = when (this) {
        PRESIDENT -> "president"; VICE_PRESIDENT -> "vicePresident"
        VICE_SCUM -> "viceScum"; SCUM -> "scum"; NEUTRAL -> "neutral"
    }
}

/** One play in the current trick. */
sealed class PresCombo {
    object Single : PresCombo()
    object Pair : PresCombo()
    object Triple : PresCombo()
    object Quad : PresCombo()
    /** Run of pairs (consecutive ranks, length 2..6 pairs). */
    data class RunOfPairs(val length: Int) : PresCombo()
}

data class PresOptions(
    /** President can announce a house rule each round (chat-enforced). */
    val allowHouseRules: Boolean = false,
    /** Display-only: revolution rule (4-of-a-kind inverts trick). */
    val revolution: Boolean = false,
)

class PresidentPlayer(val id: String, val name: String, val isHost: Boolean) {
    val hand: MutableList<PresCard> = mutableListOf()
    var rank: PresRank = PresRank.NEUTRAL
    var finished: Boolean = false
    /** 1-based finishing order. 0 if not finished. */
    var finishOrder: Int = 0
}

data class PresTrick(val combo: PresCombo, val topPower: Int, val leaderId: String)

data class PresLastPlay(
    val playerId: String,
    val cards: List<PresCard>,
    val combo: PresCombo,
)

data class PresSwap(
    val fromId: String,
    val toId: String,
    val count: Int,
    /** True if the giver picks cards (Pres-back, VP-back); false for
     *  Scum→Pres / VS→VP which give their best automatically. */
    val giverChooses: Boolean,
    var cards: List<PresCard>? = null,
)

/** Pure President / Scum / Asshole engine. */
class PresidentEngine(private val rng: Random = Random.Default) {
    val tutorialVote = com.example.ubapp.tutorials.TutorialVote()
    val players: MutableMap<String, PresidentPlayer> = linkedMapOf()
    private val seating: MutableList<String> = mutableListOf()
    val seatingSnapshot: List<String> get() = seating.toList()
    var phase: PresidentPhase = PresidentPhase.LOBBY
    var options: PresOptions = PresOptions()
        private set

    var trick: PresTrick? = null
    val passedThisTrick: MutableSet<String> = mutableSetOf()
    var lastPlay: PresLastPlay? = null
    val finishOrder: MutableList<String> = mutableListOf()
    val pendingSwaps: MutableList<PresSwap> = mutableListOf()

    var currentIndex: Int = 0
    var lastEvent: String? = null
    var roundNumber: Int = 0

    val current: PresidentPlayer? get() = if (seating.isEmpty()) null else players[seating[currentIndex]]
    val canStart: Boolean get() = phase == PresidentPhase.LOBBY && players.size in 4..7

    fun addPlayer(id: String, name: String, isHost: Boolean = false): PresidentPlayer {
        val p = PresidentPlayer(id, name, isHost); players[id] = p; return p
    }
    fun removePlayer(id: String) { if (phase == PresidentPhase.LOBBY) players.remove(id) }

    fun setOptions(o: PresOptions) {
        if (phase != PresidentPhase.LOBBY) return
        options = o
    }

    fun start() {
        if (!canStart) return
        seating.clear(); seating.addAll(players.keys.shuffled(rng))
        for (p in players.values) {
            p.rank = PresRank.NEUTRAL; p.finished = false; p.finishOrder = 0; p.hand.clear()
        }
        finishOrder.clear(); passedThisTrick.clear()
        trick = null; lastPlay = null
        roundNumber = 1
        deal()
        val leadId = seating.firstOrNull { players[it]!!.hand.contains(PresCard(PresSuit.CLUBS, 3)) }
        currentIndex = leadId?.let { seating.indexOf(it) } ?: 0
        phase = PresidentPhase.PLAYING
        lastEvent = "Round 1 — ${current!!.name} leads with 3♣"
    }

    fun startNextRound() {
        if (phase != PresidentPhase.GAME_OVER || finishOrder.isEmpty()) return
        assignRanks()
        roundNumber += 1
        for (p in players.values) { p.hand.clear(); p.finished = false; p.finishOrder = 0 }
        finishOrder.clear(); trick = null; lastPlay = null; passedThisTrick.clear()
        deal()
        pendingSwaps.clear()
        pendingSwaps.addAll(swapPlan())
        if (pendingSwaps.isEmpty()) {
            startTricksAfterSwaps()
        } else {
            phase = PresidentPhase.SWAPPING
            lastEvent = "Round $roundNumber — swap phase"
        }
    }

    fun submitSwap(fromId: String, cards: List<PresCard>): String? {
        if (phase != PresidentPhase.SWAPPING) return "not in swap phase"
        val idx = pendingSwaps.indexOfFirst { it.fromId == fromId && it.cards == null }
        if (idx < 0) return "no pending swap for you"
        val s = pendingSwaps[idx]
        val chosen: List<PresCard> = if (s.giverChooses) {
            if (cards.size != s.count)
                return "give exactly ${s.count} card${if (s.count == 1) "" else "s"}"
            cards
        } else {
            players[fromId]!!.hand.sortedByDescending { it.power }.take(s.count)
        }
        val giverHand = players[fromId]!!.hand.toMutableList()
        for (c in chosen) {
            val i = giverHand.indexOf(c)
            if (i < 0) return "card not in hand"
            giverHand.removeAt(i)
        }
        players[fromId]!!.hand.clear(); players[fromId]!!.hand.addAll(giverHand)
        players[s.toId]!!.hand.addAll(chosen)
        players[s.toId]!!.hand.sortBy { it.power }
        players[fromId]!!.hand.sortBy { it.power }
        pendingSwaps[idx] = s.copy(cards = chosen)
        lastEvent = "${players[fromId]!!.name} gave ${chosen.size} card${if (chosen.size == 1) "" else "s"} to ${players[s.toId]!!.name}"
        if (pendingSwaps.all { it.cards != null }) startTricksAfterSwaps()
        return null
    }

    private fun startTricksAfterSwaps() {
        val pres = finishOrder.firstOrNull()
        currentIndex = if (pres != null) seating.indexOf(pres).coerceAtLeast(0) else 0
        phase = PresidentPhase.PLAYING
        lastEvent = "${current!!.name} (President) leads round $roundNumber"
        pendingSwaps.clear()
        finishOrder.clear()
        for (p in players.values) { p.finished = false; p.finishOrder = 0 }
    }

    fun play(playerId: String, cards: List<PresCard>): String? {
        if (phase != PresidentPhase.PLAYING) return "not playing"
        val p = players[playerId] ?: return "unknown player"
        if (p.id != current!!.id) return "not your turn"
        if (cards.isEmpty()) return "play at least one card"
        val hand = p.hand.toMutableList()
        for (c in cards) {
            val i = hand.indexOf(c)
            if (i < 0) return "card not in hand"
            hand.removeAt(i)
        }
        val combo = classify(cards) ?: return "invalid combination"
        val power = comboPower(combo, cards)
        val t = trick
        if (t != null) {
            if (!sameType(t.combo, combo)) return "must play ${describe(t.combo)}"
            if (power <= t.topPower) return "must play higher than ${t.topPower}"
        } else if (roundNumber == 1 && lastPlay == null
                  && p.hand.contains(PresCard(PresSuit.CLUBS, 3))
                  && !cards.contains(PresCard(PresSuit.CLUBS, 3))) {
            return "first play must include 3♣"
        }
        p.hand.clear(); p.hand.addAll(hand)
        trick = PresTrick(combo, power, t?.leaderId ?: p.id)
        lastPlay = PresLastPlay(p.id, cards.toList(), combo)
        passedThisTrick.clear()
        lastEvent = "${p.name} played ${formatCards(cards)} (${describe(combo)})"
        if (p.hand.isEmpty()) {
            p.finished = true
            finishOrder.add(p.id)
            p.finishOrder = finishOrder.size
        }
        if (remainingPlayers().size <= 1) {
            for (sid in seating) {
                val pl = players[sid] ?: continue
                if (!pl.finished) {
                    pl.finished = true
                    finishOrder.add(pl.id)
                    pl.finishOrder = finishOrder.size
                }
            }
            assignRanks()
            phase = PresidentPhase.GAME_OVER
            lastEvent = "Round over"
            return null
        }
        advanceTurn()
        return null
    }

    fun pass(playerId: String): String? {
        if (phase != PresidentPhase.PLAYING) return "not playing"
        val p = players[playerId] ?: return "unknown player"
        if (p.id != current!!.id) return "not your turn"
        if (trick == null) return "lead — can't pass"
        passedThisTrick.add(p.id)
        lastEvent = "${p.name} passed"
        val lp = lastPlay
        val alive = remainingPlayers().filter { !it.finished }.map { it.id }
        if (lp != null && alive.all { it == lp.playerId || passedThisTrick.contains(it) }) {
            trick = null
            passedThisTrick.clear()
            val leader = players[lp.playerId]
            if (leader != null && !leader.finished) {
                currentIndex = seating.indexOf(lp.playerId).coerceAtLeast(0)
                lastEvent = "${leader.name} wins the trick and leads"
            } else {
                advanceTurn()
            }
            return null
        }
        advanceTurn()
        return null
    }

    fun reset() {
        phase = PresidentPhase.LOBBY
        for (p in players.values) {
            p.hand.clear(); p.rank = PresRank.NEUTRAL
            p.finished = false; p.finishOrder = 0
        }
        trick = null; lastPlay = null; passedThisTrick.clear()
        finishOrder.clear(); pendingSwaps.clear()
        currentIndex = 0; lastEvent = null; roundNumber = 0
    }

    // ----------------- internals -----------------

    private fun deal() {
        val deck = presStandardDeck().shuffled(rng)
        for ((i, c) in deck.withIndex()) {
            players[seating[i % seating.size]]!!.hand.add(c)
        }
        for (p in players.values) p.hand.sortBy { it.power }
    }

    private fun assignRanks() {
        if (finishOrder.isEmpty()) return
        val n = finishOrder.size
        for ((i, id) in finishOrder.withIndex()) {
            players[id]!!.rank = rankFor(i, n)
            players[id]!!.finishOrder = i + 1
        }
    }

    private fun rankFor(position: Int, total: Int): PresRank = when {
        position == 0 -> PresRank.PRESIDENT
        position == total - 1 -> PresRank.SCUM
        total >= 4 && position == 1 -> PresRank.VICE_PRESIDENT
        total >= 4 && position == total - 2 -> PresRank.VICE_SCUM
        else -> PresRank.NEUTRAL
    }

    private fun swapPlan(): List<PresSwap> {
        val plan = mutableListOf<PresSwap>()
        val pres = players.values.firstOrNull { it.rank == PresRank.PRESIDENT }
        val scum = players.values.firstOrNull { it.rank == PresRank.SCUM }
        val vp = players.values.firstOrNull { it.rank == PresRank.VICE_PRESIDENT }
        val vs = players.values.firstOrNull { it.rank == PresRank.VICE_SCUM }
        if (pres != null && scum != null) {
            plan.add(PresSwap(scum.id, pres.id, 2, giverChooses = false))
            plan.add(PresSwap(pres.id, scum.id, 2, giverChooses = true))
        }
        if (vp != null && vs != null) {
            plan.add(PresSwap(vs.id, vp.id, 1, giverChooses = false))
            plan.add(PresSwap(vp.id, vs.id, 1, giverChooses = true))
        }
        return plan
    }

    private fun advanceTurn() {
        val n = seating.size
        repeat(n) {
            currentIndex = (currentIndex + 1) % n
            val p = players[seating[currentIndex]]!!
            if (!p.finished) return
        }
    }

    fun remainingPlayers(): List<PresidentPlayer> =
        seating.mapNotNull { players[it] }.filter { !it.finished }

    fun classify(cards: List<PresCard>): PresCombo? {
        if (cards.isEmpty()) return null
        val counts = cards.groupingBy { it.rank }.eachCount()
        if (counts.size == 1) return when (cards.size) {
            1 -> PresCombo.Single
            2 -> PresCombo.Pair
            3 -> PresCombo.Triple
            4 -> PresCombo.Quad
            else -> null
        }
        if (counts.values.all { it == 2 } && counts.size >= 2) {
            val ranks = counts.keys.sorted()
            for (i in 1 until ranks.size) if (ranks[i] != ranks[i-1] + 1) return null
            return PresCombo.RunOfPairs(ranks.size)
        }
        return null
    }

    fun comboPower(combo: PresCombo, cards: List<PresCard>): Int =
        cards.maxOfOrNull { it.power } ?: 0

    fun sameType(a: PresCombo, b: PresCombo): Boolean = when {
        a is PresCombo.Single && b is PresCombo.Single -> true
        a is PresCombo.Pair && b is PresCombo.Pair -> true
        a is PresCombo.Triple && b is PresCombo.Triple -> true
        a is PresCombo.Quad && b is PresCombo.Quad -> true
        a is PresCombo.RunOfPairs && b is PresCombo.RunOfPairs -> a.length == b.length
        else -> false
    }

    fun describe(combo: PresCombo): String = when (combo) {
        is PresCombo.Single -> "single"
        is PresCombo.Pair -> "pair"
        is PresCombo.Triple -> "triple"
        is PresCombo.Quad -> "four of a kind"
        is PresCombo.RunOfPairs -> "${combo.length} consecutive pairs"
    }

    private fun formatCards(cards: List<PresCard>): String =
        cards.sortedBy { it.power }.joinToString(" ") { it.toString() }
}
